diff --git a/qemu_mode/build_qemu_support.sh b/qemu_mode/build_qemu_support.sh
index 118f6ebd24b4..a13c17be3656 100755
--- a/qemu_mode/build_qemu_support.sh
+++ b/qemu_mode/build_qemu_support.sh
@@ -158,6 +158,7 @@ cd qemu-$VERSION || exit 1
 
 echo "[*] Applying patches..."
 
+patch -p1 <../patches/exit-double-read.diff || exit 1
 patch -p1 <../patches/elfload.diff || exit 1
 patch -p1 <../patches/bsd-elfload.diff || exit 1
 patch -p1 <../patches/cpu-exec.diff || exit 1
@@ -192,7 +193,7 @@ if [ "$STATIC" = "1" ]; then
 	  --disable-libusb --disable-usb-redir --disable-vde --disable-vhost-net --disable-virglrenderer \
 	  --disable-virtfs --disable-vnc --disable-vte --disable-xen --disable-xen-pci-passthrough --disable-xfsctl \
 	  --enable-linux-user --disable-system --disable-blobs --disable-tools --enable-capstone=internal \
-	  --target-list="${CPU_TARGET}-linux-user" --static --disable-pie --cross-prefix=$CROSS_PREFIX || exit 1
+	  --target-list=i386-linux-user,x86_64-linux-user --static --disable-pie --cross-prefix=$CROSS_PREFIX || exit 1
 
 else
 
@@ -201,7 +202,7 @@ else
   
   ./configure --disable-system \
     --enable-linux-user --disable-gtk --disable-sdl --disable-vnc --enable-capstone=internal \
-    --target-list="${CPU_TARGET}-linux-user" --enable-pie $CROSS_PREFIX || exit 1
+    --target-list=i386-linux-user,x86_64-linux-user --enable-pie $CROSS_PREFIX || exit 1
 
 fi
 
@@ -213,14 +214,20 @@ make -j `nproc` || exit 1
 
 echo "[+] Build process successful!"
 
-echo "[*] Copying binary..."
+echo "[*] Copying binaries..."
 
-cp -f "${CPU_TARGET}-linux-user/qemu-${CPU_TARGET}" "../../afl-qemu-trace" || exit 1
+#cp -f "${CPU_TARGET}-linux-user/qemu-${CPU_TARGET}" "../../afl-qemu-trace" || exit 1
+for target in i386 x86_64; do
+  mkdir -p ../../tracers/${target} || exit 1
+  cp -f "${target}-linux-user/qemu-${target}" "../../tracers/${target}/afl-qemu-trace" || exit 1
+  cp -f "../../libradamsa.so" "../../tracers/${target}/" || exit 1
+  cp -f "../../afl-showmap" "../../tracers/${target}/" || exit 1
+done
+cp -f "${target}-linux-user/qemu-${CPU_TARGET}" "../../afl-qemu-trace" || exit 1
 
 cd ..
 ls -l ../afl-qemu-trace || exit 1
 
-echo "[+] Successfully created '../afl-qemu-trace'."
+echo "[+] Successfully build qemu mode'."
 
 if [ "$ORIG_CPU_TARGET" = "" ]; then
 
